fdbclient/SpecialKeySpace.actor.cpp

```diff
@@ -77,122 +84,108 @@ ACTOR Future<Void> SpecialKeyRangeBaseImpl::normalizeKeySelectorActor(const Spec
+// If the corresponding key is outside the whole space, it will move to the begin or the end
+// It does have overhead here since we query all keys twice in the worst case.
+// However, moving the KeySelector while handling other parameters like limits makes the code much more complex and hard
+// to maintain; Thus, separate each part to make the code easy to understand and more compact
+ACTOR Future<Void> normalizeKeySelectorActor(SpecialKeySpace* sks, Reference<ReadYourWritesTransaction> ryw,
+                                             KeySelector* ks, Optional<SpecialKeySpace::MODULE>* lastModuleRead,
+                                             int* actualOffset, Standalone<RangeResultRef>* result) {
+	state RangeMap<Key, SpecialKeyRangeBaseImpl*, KeyRangeRef>::Iterator iter =
+	    ks->offset < 1 ? sks->getImpls().rangeContainingKeyBefore(ks->getKey())
+	                   : sks->getImpls().rangeContaining(ks->getKey());
+	while ((ks->offset < 1 && iter != sks->getImpls().ranges().begin()) ||
+	       (ks->offset > 1 && iter != sks->getImpls().ranges().end())) {
+		onModuleRead(ryw, sks->getModules().rangeContaining(iter->begin())->value(), *lastModuleRead);
+		if (iter->value() != nullptr) {
+			wait(moveKeySelectorOverRangeActor(iter->value(), ryw, ks));
+		}
+		ks->offset < 1 ? --iter : ++iter;
+	}
+	*actualOffset = ks->offset;
+	if (iter == sks->getImpls().ranges().begin())
+		ks->setKey(sks->getKeyRange().begin);
+	else if (iter == sks->getImpls().ranges().end())
+		ks->setKey(sks->getKeyRange().end);
+
+	if (!ks->isFirstGreaterOrEqual()) {
+		// The Key Selector points to key outside the whole special key space
+		TraceEvent(SevInfo, "KeySelectorPointsOutside")
+		    .detail("TerminateKey", ks->getKey())
+		    .detail("TerminateOffset", ks->offset);
+		if (ks->offset < 1 && iter == sks->getImpls().ranges().begin())
+			result->readToBegin = true;
+		else
+			result->readThroughEnd = true;
+		ks->offset = 1;
+	}
+	return Void();
+}
+
+ACTOR Future<Standalone<RangeResultRef>> SpecialKeySpace::checkModuleFound(SpecialKeySpace* sks,
                                                                            Reference<ReadYourWritesTransaction> ryw,
                                                                            KeySelector begin, KeySelector end,
                                                                            GetRangeLimits limits, bool reverse) {
-	std::pair<Standalone<RangeResultRef>, Optional<SpecialKeyRangeBaseImpl*>> result =
-	    wait(SpecialKeySpace::getRangeAggregationActor(pks, ryw, begin, end, limits, reverse));
+	std::pair<Standalone<RangeResultRef>, Optional<SpecialKeySpace::MODULE>> result =
+	    wait(SpecialKeySpace::getRangeAggregationActor(sks, ryw, begin, end, limits, reverse));
 	if (ryw && !ryw->specialKeySpaceRelaxed()) {
-		auto* module = result.second.orDefault(nullptr);
-		if (module == nullptr) {
+		auto module = result.second;
+		if (!module.present()) {
```

Comment:
Wouldn't we want to throw if `module.get() == SpecialKeySpace::MODULE::UNKNOWN` ?

Reply:
Nice catch!
My previous test did not catch this since all tests using keys not `KeySelector` do not have chance to change this `module`.
However, if one `keySelector` points to the last module and also has `offset > 1`, it will change the `module` to `UNKNOWN` and the `no_found_error` is missed.
 I have added the test to catch this bug and fixed it.